superset/common/query_context_processor.py

```diff
@@ -307,6 +324,40 @@ def _get_timestamp_format(
 
         return df
 
+    @staticmethod
+    def get_time_grain(
+        query_context: QueryContext, query_object: QueryObject
+    ) -> Any | None:
+        if query_context.form_data:
+            return query_context.form_data.get("time_grain_sqla")
```

Comment:
I think this should be avoided. The form data has been introduced fairly recently in query_context, and isn't really meant to influence rendering of the query_object (IIRC, it was originally added to relay some context about query timeouts that are defined in the chart object). Is there some reason we are primarily looking at the form data here, and not the query object?

Reply:
I can remove this line given that the created function deals with the different `query_object` variations that depend on `GENERIC_CHART_AXES`. I was using the `form_data` because it was the only structure that looks consistent independently of the feature flag.

Reply:
I understand the confusion this causes. I'll try to prepare a follow-up that moves the relevant info from form data into query context/object and removes form data all together.